### PR TITLE
feat: vault hub solvency tests

### PR DESCRIPTION
**Pull request by kovalgek**
A short summary of the changes.

## Context

Create solvency tests for VaultHub

## Problem

The VaultHub with connected vaults has a complex state. Solvency tests can help to find issues (if they exist) by repeating many operations with random parameteres.  

## Solution

These tests create and connect Vault to the VaultHub in a random time and then call mint/burn/rebalance/receive_rewards/receive_penalty operations on them. At the same time simulate core protocol staking and rewards mechanism.

### CODE REVIEW

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+    uint256 private connectedVaults;
+    uint256 private disconnectedVaults;
+    uint256 private relativeShareLimitBp;
+
+    function deploy(uint256 _seed) public {
+        rnd.seed = _seed;
+
+        depositContract = new DepositContract__MockForStakingVault();
+
+        lido = new LidoMock(7810237 * 10 ** 18, 9365361 * 10 ** 18, 0);
```
Comment (tamtamchik) [inline test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: Need a comment about magic numbers

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+        LidoLocatorMock lidoLocator = new LidoLocatorMock(depositor, accounting, treasury);
```
Comment (tamtamchik) [inline test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: Why do we need a mock here? Can't we use the real one?

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol [hunk 3]
```diff
@@ -0,0 +1,804 @@
...
+        bytes32 vaultMasterRole = vaultHubProxy.VAULT_MASTER_ROLE();
+        vm.prank(owner);
+        vaultHubProxy.grantRole(vaultMasterRole, owner);
+
+        bytes32 vaultRegistryRole = vaultHubProxy.VAULT_REGISTRY_ROLE();
+        vm.prank(owner);
+        vaultHubProxy.grantRole(vaultRegistryRole, owner);
+
+        rewardSimulatorForValidator = new RewardSimulator(_seed, 300, 400, 32 ether);
+        rewardSimulatorForCoreProtocol = new RewardSimulator(_seed, 100, 200, lido.getTotalPooledEther());
```
Comment (tamtamchik) [inline test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: 🤔 maybe add comments to highlight why vaults perform better than core?

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol
Comment (tamtamchik) [inline test/0.8.25/vaults/staking-vault/VaultHub.t.sol:1]: VaultHub.t.sol ⇒ VaultHubSolvency.t.sol

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+        vm.prank(owner);
+        vaultHubProxy.mintShares(address(_stakingVault), address(owner), amountOfSharesToMint);
+        console2.log("Mint random shares", address(_stakingVault), amountOfSharesToMint);
+
+        totalSharesMinted += amountOfSharesToMint;
+    }
+
+    function transitionRandomBurn(StakingVault _stakingVault) internal {
+        VaultHub.VaultSocket memory socket = vaultHubProxy.vaultSocket(address(_stakingVault));
+        uint256 amountOfSharesToBurn = rnd.randAmountD18();
```
Comment (tamtamchik) [inline test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: Maybe use cap for burning, so it'll burn every time

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+            console2.log("Receive random reward", address(_stakingVault), dailyReward);
+        }
+    }
+
+    function transitionVaultRandomPenalty(StakingVault _stakingVault) internal {
+        VaultHub.VaultSocket memory socket = vaultHubProxy.vaultSocket(address(_stakingVault));
+
+        uint256 valuationThreshold = (_stakingVault.valuation() * (TOTAL_BASIS_POINTS - socket.rebalanceThresholdBP)) /
+            TOTAL_BASIS_POINTS;
+        if (valuationThreshold < lido.getPooledEthByShares(socket.sharesMinted)) {
```
Comment (tamtamchik) [inline test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: We need to allow it to go to the unhealthy state and then, on the next interaction, force rebalance and disconnect.

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+        uint256 sharesLeftover = 0;
+        for (uint256 i = 0; i < vaults.length; i++) {
+            Vault memory vault = vaults[i];
+            sharesLeftover += vaultHubProxy.vaultSocket(address(vault.stakingVaultProxy)).sharesMinted;
+        }
+        console2.log("Total shares minted", totalSharesMinted);
+        console2.log("Total shares burned", totalSharesBurned);
+        console2.log("Shares leftover", sharesLeftover);
+
+        assertEq(totalSharesMinted, totalSharesBurned + sharesLeftover);
```
Comment (tamtamchik) [inline test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: 🤔  maybe assert vault valuation is enough to cover minted shares

**Review by tamtamchik — COMMENTED** [review]
👀 🔥 💪

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol (hunk 2 above)
Comment (kovalgek) [reply test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: The real object includes many members and extra methods, but we only need three of them, so using a mock here is, in my view, the simplest solution.

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol (hunk 3 above)
Comment (kovalgek) [reply test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: I've changed these numbers to be more realistic.

File: test/0.8.25/vaults/staking-vault/VaultHub.t.sol (hunk 5 above)
Comment (kovalgek) [reply test/0.8.25/vaults/staking-vault/VaultHub.t.sol]: can't do this because forceRebalance function reverts if vault is in a bad dept state

File: test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol [hunk 7]
```diff
@@ -4,15 +4,15 @@
...
 //                                       inOutDelta += x  inOutDelta -= x
 //                                              │              │
 //                                              │              │
-//   ┌──────────────────┐──report(old data)──▶┌─▼──────────────▼─┐    depositToBeaconChain   ┌───────────────────────┐
-//   │                  │                     │                  │◀───SV.balance -= deposit──│       Depositor       │
+//   ┌──────────────────┐──report(old data)──▶┌─▼──────────────┴─┐   depositToBeaconChain    ┌───────────────────────┐
+//   │                  │                     │                  │───SV.balance -= deposit──▶│       Depositor       │
 //   │                  │     rebalance       │                  │                           └───────────────────────┘
 //   │     VaultHub     │─  SV.balance -= x──▶│   StakingVault   │
 //   │                  │   inOutDelta -= x   │                  │           rewards         ┌───────────────────────┐
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol]: maybe we should use "pragma solidity 0.8.25;" here?

File: test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol [hunk 8]
```diff
@@ -27,126 +27,103 @@ import {Math} from "@openzeppelin/contracts-v5.2/utils/math/Math.sol";
...
+    function predepositGuarantee() public view returns (address) {
+        return depositor;
+    }
+}
+
+contract VaultHubMock is RewardSimulator {
+    event Mock__Rebalanced(address indexed vault, uint256 amount);
 
-contract VaultHubMock is Test {
     using RandomLib for RandomLib.Storage;
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol]: redundant library usage

File: test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol [hunk 9]
```diff
@@ -212,19 +190,25 @@ contract StakingVaultTest is Test {
...
+
+        vm.prank(owner);
+        stakingVaultProxy.authorizeLidoVaultHub();
     }
 
     function performMajorStateTransition(bool _hasValidator, bool _isConnectedToHub) internal {
+        vm.warp(block.timestamp + rnd.randInt(SECONDS_PER_DAY));
+
         bool oldIsConnectedToHub = _isConnectedToHub;
         bool oldHasValidator = _hasValidator;
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol]: I guess we can just read it (_isConnectedToHub, _hasValidator) from storage instead of receiving it as a function parameter.

File: test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol [hunk 10]
```diff
@@ -249,7 +233,7 @@ contract StakingVaultTest is Test {
     }
 
     function randomTransition(bool _hasValidator, bool _isConnectedToHub) internal {
-        vm.warp(block.timestamp + rnd.randInt(2 * SECONDS_PER_DAY));
+        vm.warp(block.timestamp + rnd.randInt(SECONDS_PER_DAY));
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol]: Why did you decide to use a 1-day range instead of 2 days?

File: test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol [hunk 11]
```diff
@@ -460,81 +450,108 @@ contract StakingVaultTest is Test {
...
     function transitionDisconnectVaultFromHub() internal {
         console2.log("------Disconnect Vault from Hub------");
 
-        uint256 valuation = stakingVaultProxy.valuation();
+        uint256 valuation = stakingVaultProxy.totalValue();
         int256 inOutDelta = stakingVaultProxy.inOutDelta();
 
         vm.prank(address(vaultHub));
-        stakingVaultProxy.report(valuation, inOutDelta, 0);
+        stakingVaultProxy.report(uint64(block.timestamp), valuation, inOutDelta, 0);
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol]: I'm not sure it's necessary, but we could add `deauthorizeLidoVaultHub` and `resetLocked` calls to additionally test these vault states.

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 12]
```diff
@@ -0,0 +1,718 @@
...
+
+    function testSolvencyBadPerformingValidators() external {
+        runTests(123618273619736182376, TestMode.BadPerformingValidators);
+    }
+
+    function testSolvencyWellPerformingValidators() external {
+        runTests(23172389139823, TestMode.WellPerformingValidators);
+    }
+
+    function testFuzz_SolvencyAllTransitions(uint256 _seed) external {
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol:611]: By default forge runs 256 iterations for fuzz tests. I think we can move this value into a config, so that we can run more iterations - for example in CI to improve reliability (maybe in a separate pre-release stage). But that’s just an idea)

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 13]
```diff
@@ -0,0 +1,718 @@
...
+        totalShares -= _amountOfShares;
+    }
+}
+
+contract VaultHubTest is Test {
+    using RandomLib for RandomLib.Storage;
+
+    struct Vault {
+        StakingVault stakingVaultProxy;
+        bool isValidatorPerformingWell;
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: Unused variable

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 14]
```diff
@@ -0,0 +1,718 @@
...
+        deploy(_seed);
+
+        createAndConnectVault(true, _testMode);
+
+        for (uint256 iterationIdx = 0; iterationIdx < ITERATIONS; iterationIdx++) {
+            if (vaults.length < CONNECTED_VAULTS_LIMIT && rnd.randInt(100) < CHANCE_TO_CREATE_AND_CONNECT_VAULT) {
+                createAndConnectVault(false, _testMode);
+            }
+
+            vm.warp(block.timestamp + SECONDS_PER_DAY / 2);
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: Why is the range limited to half a day here? It seems like testing a case where more than one day passes could also be interesting.

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 15]
```diff
@@ -0,0 +1,718 @@
...
+        for (uint256 i = 0; i < vaults.length; i++) {
+            Vault memory vault = vaults[i];
+            uint256 sharesMinted = vaultHubProxy.vaultSocket(address(vault.stakingVaultProxy)).liabilityShares;
+
+            if (getVaultState(vault) != VaultState.BadDebt) {
+                assertLe(lido.getPooledEthBySharesRoundUp(sharesMinted), vault.stakingVaultProxy.totalValue());
+            }
+
+            uint256 relativeMaxShareLimitPerVault = (lido.getTotalShares() * relativeShareLimitBp) / TOTAL_BASIS_POINTS;
+            assertLe(sharesMinted, relativeMaxShareLimitPerVault);
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: It seems like this check should apply to the total `sharesMinted` across all vaults.

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 16]
```diff
@@ -0,0 +1,718 @@
...
+
+        // average validator APR is between 2.8-5.7%
+        rewardSimulatorForValidator = new RewardSimulator(_seed, 280, 570, 32 ether);
+        // average core protocol APR is between 2.78-3.6%
+        rewardSimulatorForCoreProtocol = new RewardSimulator(_seed, 278, 360, lido.getTotalPooledEther());
+
+        uint256 shareLimit = 35 ether + rnd.randInt(5) * 1 ether; // between 35 and 40 ETH
+        uint256 reserveRatioBp = 1000 + rnd.randInt(1000); // between 10% and 20%
+        uint256 forcedRebalanceThresholdBp = 500 + rnd.randInt(reserveRatioBp - 500); // between 5% and 95% of reserveRatioBp
+        uint256 treasuryFeeBp = 500 + rnd.randInt(500); // between 1% and 10%
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]:
```suggestion
        uint256 treasuryFeeBp = 500 + rnd.randInt(500); // between 5% and 10%
```

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 17]
```diff
@@ -0,0 +1,718 @@
...
+            );
+        }
+
+        bytes32[] memory paddedData = padDataIfNeeded(data);
+        bytes32 root = merkle.getRoot(paddedData);
+
+        vm.prank(accounting);
+        vaultHubProxy.updateReportData(uint64(block.timestamp), root, "test-cid");
+
+        for (uint256 i = 0; i < vaults.length; i++) {
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: I'm not sure, but it might be interesting if some vaults submitted lazy reports immediately while others didn't — to test those scenarios as well.

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 18]
```diff
@@ -0,0 +1,718 @@
...
+            }
+
+            if (state == VaultState.MintingAllowed) {
+                transitionMintMaxAllowedShares(vault.stakingVaultProxy);
+            } else if (state == VaultState.Healthy || state == VaultState.Unhealthy) {
+                if (rnd.randInt(100) < CHANCE_TO_CALL_REBALANCE_OR_BURN) {
+                    if (rnd.randBool()) {
+                        transitionRandomBurn(vault.stakingVaultProxy);
+                    } else {
+                        transitionRandomRebalance(vault.stakingVaultProxy);
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: It seems like rebalance shouldn't or can't be called for Healthy vaults — maybe we should move the Healthy case into a separate if branch.

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 19]
```diff
@@ -0,0 +1,718 @@
...
+    struct Vault {
+        StakingVault stakingVaultProxy;
+        bool isValidatorPerformingWell;
+        ValidatorState validatorState;
+        uint256 lifetime;
+        uint256 lastInactivePenaltyTime;
+        VaultState lastState;
+    }
+
+    enum VaultAction {
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: unused enum

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 20]
```diff
@@ -0,0 +1,718 @@
...
+    address private accounting = makeAddr("accounting");
+    address private treasury = makeAddr("treasury");
+    address private depositor = makeAddr("depositor");
+    address private nodeOperator = makeAddr("nodeOperator");
+
+    RandomLib.Storage private rnd;
+
+    uint256 internal constant ITERATIONS = 170;
+    uint256 internal constant CONNECTED_VAULTS_LIMIT = 100;
+    uint256 internal constant BAD_BEHAVIOUR_ITERATIONS_LIMIT = 10;
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: unused constant

File: test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol [hunk 21]
```diff
@@ -0,0 +1,718 @@
...
+        vm.prank(owner);
+        vaultHubProxy.mintShares(address(_stakingVault), address(owner), amountOfSharesToMint);
+        console2.log("Mint random shares", address(_stakingVault), amountOfSharesToMint);
+
+        totalSharesMinted += amountOfSharesToMint;
+    }
+
+    function transitionRandomBurn(StakingVault _stakingVault) internal {
+        VaultHub.VaultSocket memory socket = vaultHubProxy.vaultSocket(address(_stakingVault));
+        uint256 amountOfSharesToBurn = rnd.randInt(socket.liabilityShares);
```
Comment (dry914) [inline test/0.8.25/vaults/staking-vault/VaultHubSolvency.t.sol]: Here, edge cases are especially important — it makes sense to assign them higher probability during randomization. For instance, a 10% chance of selecting 0 and another 10% for the full `socket.liabilityShares`.

**Review by dry914 — COMMENTED** [review]
Overall, it looks solid — great job!

File: test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol (hunk 9 above)
Comment (kovalgek) [reply test/0.8.25/vaults/staking-vault/StakingVaultSolvency.t.sol]: Not valid anymore
